text/0003-templates.md

```diff
@@ -0,0 +1,393 @@
+  "version": "0.1.0",
+  "pnpm": {
+    "template": "authoring"
+  },
+  "author": "Example Organization <team@organization.example>",
+  "license": "MIT",
+  "repository": "git@organization.example/example/project"
+}
+```
+
+A package referencing the `@example/authoring-template` above will have the following on-disk and in-memory representations.
+
+**On-Disk**
+
+```json5
+{
+  "name": "@example/react-components",
+  "pnpm": {
+    "templates": {
+      "authoring": ["@example/authoring-template@0.1.0"],
+    }
+  }
+}
+```
+
+**In-Memory and Publish Time**
+
+```json5
+{
+  "name": "@example/react-components",
+  "author": "Example Organization <team@organization.example>",
+  "license": "MIT",
+  "repository": "git@organization.example/example/project"
+}
+```
+
+### Toolkit
+
+The `pnpm.templates.toolkit` flavor copies the `dependencies` block of the template and spreads it into the `devDependencies` of the referencing package.
+
+Given the following template:
+
+```json5
+{
+  "name": "@example/react-toolkit",
+  "dependencies": {
+    "jest": "^29.4.3",
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+    "redux": "^4.2.1"
+  }
+}
+```
+
+A package referencing the `@example/react-toolkit` above will have the following on-disk and in-memory representations.
+
+**On-Disk**
+
+```json5
+{
+  "name": "@example/react-components",
+  "pnpm": {
+    "templates": {
+      "toolkit": ["@example/react-toolkit@0.1.0"],
+    }
+  }
+}
+```
+
+**In-Memory and Publish Time**
+
+```json5
+{
+  "name": "@example/react-components",
+  "dependencies": {
+    "redux": "^4.2.1"
+  },
+  "devDependencies": {
+    "jest": "^29.4.3",
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+    "redux": "^4.2.1"
+  }
+}
+```
+
+This allows package authors to share a reusable set of dependencies that can be imported or required in the referencing package. Dependencies of dependencies are not normally visible in this manner since pnpm sets up a semi-strict `node_modules` structure by default.
+
+### Catalog
+
+The `pnpm.templates.catalog` flavor allows packages to declare a dependency using a version specifier from the referenced template. The `dependencies` block of the template will be available to reference through the `catalog:` version specifier protocol.
+
+```json5
+{
+  "name": "@example/frontend-catalog",
+  "pnpm": {
+    "template": "catalog"
+  },
+  "dependencies": {
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+    "redux": "^4.2.0",
+    "react-redux": "^8.0.0"
+  }
+}
+```
+
+A package referencing the `@example/frontend-catalog` above will have the following on-disk and in-memory representations.
+
+**On-Disk**
+
+```json5
+{
+  "name": "@example/react-components",
+  "pnpm": {
+    "templates": {
+      "catalog": {
+        "example-frontend": "@example/frontend-catalog@0.1.0"
+      }
+    }
+  },
+  "dependencies": {
+    "react": "catalog:example-frontend",
+    "redux": "catalog:example-frontend"
```

Comment:
do we need this tagging when there is only one catalog?
can't it be just
```suggestion
    "redux": "catalog:"
```

Even if there are multiple catalogs, they will probably not contain the same dependencies in most cases.

Reply:
Yeah, I could use your opinion on the right level of explicitness to require and when.

My thought process for requiring `catalog:<name>` even when only 1 catalog is configured would be to avoid pain/friction on maintainers as their projects grow.

1. **1 catalog**: I suspect most projects will indeed start out with one catalog. Suppose we allow just `catalog:` without a name in this case.
2. **Multiple catalogs with no overlap**: When users start using multiple catalogs, they're technically fine as long as there's no overlap. They can continue to use `catalog:` without a name.
3. **Multiple catalogs with overlap**: Now we have to start emitting errors about `catalog:` resolution being ambiguous. Users have to go back to existing `catalog:` specifiers and figure out which catalog was originally intended for reference.

I think there's several particularly frustrating variants of the last scenario:

- If a user attempts to upgrade their external catalog versions, but one of the updated catalogs introduces a new dependency that creates an overlap. I don't think we want to add potential friction to the upgrade process.
- If there's already many catalogs in use (suppose 5 or more), then the user introducing a new catalog with an ambiguity has to go through all conflicting `catalog:` specifiers and determine which of the many (5+) catalogs they need to specify.

The idea of requiring `catalog:<name>` was that it's fairly low upfront friction and allows projects to grow from 1 -> 3 with zero continued friction.

> Even if there are multiple catalogs, they will probably not contain the same dependencies in most cases.

I agree most catalog users won't see overlap. I think it's a matter of how much we want to avoid the "surprise" of frustrating users by forcing them to go back and clarify existing `catalog:` specifiers when they do hit this case.

Reply:
There's a possible middle ground where we begin requiring `catalog:<name>` as soon as 2 catalogs are used, instead of later down the line when multiple catalogs create overlap.

But I'm worried that introduces enough friction that users will just avoid adding the second catalog.

Reply:
You wouldn't use specs for the same dependency from different catalogs, so it is OK if this is not supported. For instance, we can place catalogs into an array and the last one would override previous ones.


```json
{
  "pnpm": {
    "templates": {
      "catalog": [
        "@example/frontend-catalog@0.1.0",
        "@example/backend-catalog@0.1.0"
      ]
    }
  }
}
```

If both `frontent-catalog` and `backend-catalog` have `react`, then `react` of `backend-catalog` will be used.

Reply:
That seems okay for the smaller (~2) catalog case. Suppose 3 catalogs were configured.

```json5
{
  "pnpm": {
    "templates": {
      "catalog": [
        "@example/frontend-catalog@0.1.0",
        "@example/backend-catalog@0.1.0",
        "@other/some-backend-framework@0.1.0",
      ]
    }
  },
  "dependencies": {
    "example-backend-util-lib": "catalog:" // ← Resolved from @example/backend-catalog
  }
}
```

But then `@other/some-backend-framework@0.2.0` adds a `example-backend-util-lib`. If I want to continue using the spec from `@example/backend-catalog` for the utility package, how would I do that?

I suppose with the array syntax, we could allow:

```json5
{
  "pnpm": {
    "templates": {
      "catalog": [
        "@example/frontend-catalog@0.1.0",
        "@example/backend-catalog@0.1.0",
        "@other/some-backend-framework@0.2.0",
      ]
    }
  },
  "dependencies": {
    "example-backend-util-lib": "catalog:@example/backend-catalog"
  }
}
```

But I think I currently lean towards always requiring the name to avoid these "gotchas" down the line. Open to other ideas.

Reply:
> You wouldn't use specs for the same dependency from different catalogs, so it is OK if this is not supported.

For completely internal catalogs maintained by the same project owner, I'd agree that they should be designed well enough that they wouldn't overlap.

I'm mostly worried about a mix of internal and external catalogs. An external catalog outside of the project maintainer's control could add new dependencies in an update that cause ambiguity/conflicts that would be hard to audit.

Reply:
It will be quite verbose but I don't have better ideas.

Reply:
at least make it possible to define a default catalog.

Reply:
How do we feel about `default` as a special name? The idea would be to make `catalog:` simply a shorthand for `catalog:default`.

```json
{
  "pnpm": {
    "templates": {
      "catalog": {
        "default": "@example/frontend-catalog@0.1.0",
        "backend": "@example/backend-catalog@0.1.0",
        "other": "@external/some-backend-framework@0.2.0"
      }
    }
  },
  "dependencies": {
    "react": "catalog:"
    "example-backend-util-lib": "catalog:backend"
  }
}
```

Alternatively, if we're okay introducing an additional shorthand syntax to make aliases optional:

```json
{
  "pnpm": {
    "templates": {
      "catalog": {
        "default": "@example/frontend-catalog@0.1.0",
        "@example/backend-catalog": "0.1.0",
        "@external/some-backend-framework": "0.2.0"
      }
    }
  },
  "dependencies": {
    "react": "catalog:"
    "example-backend-util-lib": "catalog:@example/backend-catalog"
  }
}
```

Reply:
looks good